pydantic/_internal/_model_construction.py

```diff
@@ -88,6 +88,7 @@ def __new__(
             base_field_names, class_vars, base_private_attributes = mcs._collect_bases_data(bases)
 
             config_wrapper = ConfigWrapper.for_model(bases, namespace, kwargs)
+            cls_name = config_wrapper.config_dict.get('model_name') or cls_name
```

Comment:
Would it be possible to have some kind of "backup" of the original class name if overridden by `config.model_name`? I'm making use of `cls.__name__` [here](https://github.com/pydantic/pydantic/pull/6563/files#diff-515831ae80b48a711193d4896331a746d0e6e2ff0116621b58a068e4d242acc2R70), and comparing it to AST-parsed file.

In fact it would be great to have this PR merged before https://github.com/pydantic/pydantic/pull/6563/, I will then be able to rebase on top of it and add this "backup" field

Reply:
I think that if `cls.__name__` is used somewhere else, maybe it would be better to keep its original behaviour and create an extra attribute for the model_name that will explicitly interact with the `v1.schema` functions. What do you think?

Reply:
Yes this seems better: `__name__` is a Python implementation detail and people should expect it to be the same as defined in the source code (it might also be used by documentation tools). Perhaps you can define a specific attribute that will be used for schema generation